src/main/host/timer.rs

```diff
@@ -0,0 +1,237 @@
+use std::sync::{Arc, Weak};
+
+use atomic_refcell::AtomicRefCell;
+use log::trace;
+
+use crate::core::support::emulated_time::EmulatedTime;
+use crate::core::support::simulation_time::SimulationTime;
+use crate::core::work::task::TaskRef;
+use crate::core::worker::Worker;
+use crate::utility::{Magic, ObjectCounter};
+
+use super::host::Host;
+
+pub struct Timer {
+    magic: Magic<2537832161>,
+    _counter: ObjectCounter,
+    // Internals in an Arc so that we can schedule tasks that refer back to it.
+    // This is the only persistent strong reference - callbacks use a Weak
+    // reference.  i.e. dropping the outer object will drop this field as well;
+    // scheduled callbacks with weak references that can't be upgraded become
+    // no-ops.
+    internal: Arc<AtomicRefCell<TimerInternal>>,
+}
+
+struct TimerInternal {
+    next_expire_time: Option<EmulatedTime>,
+    expire_interval: SimulationTime,
+    expiration_count: u64,
+    next_expire_id: u64,
+    min_valid_expire_id: u64,
+    on_expire: Box<dyn Fn(&mut Host) + Send + Sync>,
+}
+
+impl Timer {
+    /// Create a new Timer that synchronously executes `on_expire` on
+    /// expiration.  `on_expire` will cause a panic if it calls mutable methods
```

Comment:
What does "synchronously" mean in this context?

Reply:
I meant as opposed to pushing it to the scheduler queue. Changed to 'directly'; I think that's a little clearer.